File: frontend/src/types/practice.ts
export type QuestionType = "vocabulary_recall" | "character_recognition";

export type KnowledgeStatus = "new" | "learning" | "known";

export type RecallDirection =
  | "meaning_to_hanzi"
  | "hanzi_to_meaning"
  | "pinyin_to_hanzi";

export type RecognitionDirection =
  | "character_to_meaning"
  | "character_to_pinyin"
  | "meaning_to_character"
  | "pinyin_to_character";

export interface AnswerOption {
  text: string;
  is_correct: boolean;
}

export interface Question {
  id: string;
  type: QuestionType;
  order: number;
  prompt: string;
  correct_answers: string[];
  vocabulary_id: string | null;
  character: string | null;
  options: AnswerOption[];
  is_multiple_choice: boolean;
}

export interface Exercise {
  id: string;
  learner_id: string;
  type: QuestionType;
  status: string;
  questions: Question[];
  category_id: string | null;
  knowledge_status_filter: string | null;
  question_count: number;
  candidate_count: number;
  created_at: string;
}

export interface GenerateVocabularyRecallRequest {
  count?: number;
  category_id?: string | null;
  knowledge_status?: KnowledgeStatus | null;
  direction?: RecallDirection;
}

export interface GenerateCharacterRecognitionRequest {
  count?: number;
  knowledge_status?: KnowledgeStatus | null;
  direction?: RecognitionDirection;
}

export interface SubmitAnswerRequest {
  exercise_id: string;
  question_id: string;
  question_type: QuestionType;
  raw_answer: string;
  correct_answers: string[];
  vocabulary_id?: string | null;
  character?: string | null;
  response_time_ms?: number | null;
}

export interface SubmitAnswerResponse {
  attempt_id: string;
  is_correct: boolean;
  previous_status: string | null;
  new_status: string | null;
  raw_answer: string;
  response_time_ms: number | null;
}

File: backend/src/chinese_learning/presentation/rest/schemas/practice.py
"""Request/response schemas for practice endpoints."""

from pydantic import BaseModel, Field


class AnswerOptionSchema(BaseModel):
    text: str
    is_correct: bool


class QuestionSchema(BaseModel):
    id: str
    type: str
    order: int
    prompt: str
    correct_answers: list[str]
    vocabulary_id: str | None = None
    character: str | None = None
    options: list[AnswerOptionSchema] = Field(default_factory=list)
    is_multiple_choice: bool = False


class ExerciseSchema(BaseModel):
    id: str
    learner_id: str
    type: str
    status: str
    questions: list[QuestionSchema]
    category_id: str | None = None
    knowledge_status_filter: str | None = None
    question_count: int
    candidate_count: int
    created_at: str


class GenerateVocabularyRecallRequest(BaseModel):
    count: int = Field(default=10, ge=1, le=50)
    category_id: str | None = None
    knowledge_status: str | None = Field(
        default=None,
        description="new | learning | known",
    )
    direction: str = Field(default="meaning_to_hanzi")


class GenerateCharacterRecognitionRequest(BaseModel):
    count: int = Field(default=10, ge=1, le=50)
    knowledge_status: str | None = None
    direction: str = Field(default="character_to_meaning")


class SubmitAnswerRequest(BaseModel):
    exercise_id: str
    question_id: str
    question_type: str
    raw_answer: str = Field(..., min_length=0, max_length=500)
    correct_answers: list[str] = Field(..., min_length=1)
    vocabulary_id: str | None = None
    character: str | None = None
    response_time_ms: int | None = Field(default=None, ge=0)


class SubmitAnswerResponse(BaseModel):
    attempt_id: str
    is_correct: bool
    previous_status: str | None
    new_status: str | None
    raw_answer: str
    response_time_ms: int | None = None
